File: wasm/src/localemethods.cpp
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "wasm/localemethods.h"
#include "apl/apl.h"
#include "apl/utils/log.h"

namespace apl {
    namespace wasm {
        WasmLocaleMethods::WasmLocaleMethods(emscripten::val toUpperCaseCallback,
                                             emscripten::val toLowerCaseCallback)
                : apl::LocaleMethods()
                , toUpperCaseCallback(toUpperCaseCallback)
                , toLowerCaseCallback(toLowerCaseCallback) {}

        std::string
        WasmLocaleMethods::toUpperCase(const std::string &value, const std::string &locale) {
            if (toUpperCaseCallback.isUndefined()) {
                LOG(LogLevel::ERROR) << "There is no toUpperCaseCallback";
                return value;
            }
            return toUpperCaseCallback(value, locale).as<std::string>();
        }

        std::string
        WasmLocaleMethods::toLowerCase(const std::string &value, const std::string &locale) {
            if (toLowerCaseCallback.isUndefined()) {
                LOG(LogLevel::ERROR) << "There is no toLowerCaseCallback";
                return value;
            }
            return toLowerCaseCallback(value, locale).as<std::string>();
        }

    } // namespace wasm
} // namespace apl

File: js/apl-html/src/utils/LocaleMethods.ts
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

function toUpperCase(value : string, locale : string) : string {
    return callSafely(value, locale, 'toLocaleUpperCase');
}

function toLowerCase(value : string, locale : string) : string {
    return callSafely(value, locale, 'toLocaleLowerCase');
}

function callSafely(value : string, locale : string, localeFunctionName : string) {
    if (!locale) {
        return value[localeFunctionName]();
    }
    return value[localeFunctionName](locale);
}

export const LocaleMethods = {
    toUpperCase,
    toLowerCase
};
